// @spartan-ng/cli/src/generators/ui/primivite-deps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDependentPrimitives = exports.primitiveDependencies = void 0;
exports.primitiveDependencies = {
    accordion: ['icon'],
    alert: ['icon'],
    alertdialog: ['button'],
    aspectratio: [],
    avatar: [],
    badge: [],
    breadcrumb: ['icon'],
    button: [],
    calendar: ['button', 'icon'],
    card: [],
    carousel: [],
    checkbox: ['icon'],
    collapsible: [],
    command: ['button', 'icon'],
    contextmenu: [],
    datepicker: ['calendar', 'icon', 'popover'],
    dialog: [],
    formfield: [],
    hovercard: [],
    icon: [],
    input: ['formfield'],
    label: [],
    menu: ['icon'],
    menubar: [],
    pagination: ['button', 'icon'],
    popover: [],
    progress: [],
    radiogroup: [],
    scrollarea: [],
    select: ['icon', 'formfield'],
    separator: [],
    sheet: [],
    skeleton: [],
    sonner: ['icon'],
    spinner: [],
    switch: [],
    table: [],
    tabs: [],
    toggle: [],
    tooltip: [],
};
const getDependentPrimitives = (primitives) => {
    const dependentPrimitives = new Set();
    for (const primitive of primitives) {
        const deps = exports.primitiveDependencies[primitive] ?? [];
        for (const dep of deps) {
            if (!primitives.includes(dep)) {
                // only add the dependent primitive if it's not already in the list of primitives to create
                dependentPrimitives.add(dep);
            }
        }
    }
    return Array.from(dependentPrimitives);
};
exports.getDependentPrimitives = getDependentPrimitives;
