// @spartan-ng/cli/src/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOrCreateConfig = getOrCreateConfig;
const devkit_1 = require("@nx/devkit");
const enquirer_1 = require("enquirer");
const configPath = 'components.json';
async function getOrCreateConfig(tree, defaults) {
    if (tree.exists(configPath)) {
        return (0, devkit_1.readJson)(tree, configPath); // TODO: Parse with zod and handle errors
    }
    console.log('Configuration file not found, creating a new one...');
    const { componentsPath } = (await (0, enquirer_1.prompt)([
        {
            type: 'input',
            required: true,
            name: 'componentsPath',
            message: 'Choose a directory to place your spartan libraries, e.g. libs/ui',
            initial: defaults?.componentsPath ?? 'libs/ui',
            skip: !!defaults?.componentsPath,
        },
    ]));
    const config = { componentsPath };
    tree.write(configPath, JSON.stringify(config, null, 2));
    return config;
}
