// @spartan-ng/brain/fesm2022/spartan-ng-brain-switch.mjs
import * as i0 from '@angular/core';
import { Component, forwardRef, inject, DestroyRef, Renderer2, PLATFORM_ID, ElementRef, ChangeDetectorRef, signal, model, input, linkedSignal, booleanAttribute, output, viewChild, computed, effect, ViewEncapsulation, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { isPlatformBrowser, DOCUMENT } from '@angular/common';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

class BrnSwitchThumbComponent {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchThumbComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "19.2.8", type: BrnSwitchThumbComponent, isStandalone: true, selector: "brn-switch-thumb", host: { attributes: { "role": "presentation" }, listeners: { "click": "$event.preventDefault()" } }, ngImport: i0, template: '', isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchThumbComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-switch-thumb',
                    template: '',
                    host: {
                        role: 'presentation',
                        '(click)': '$event.preventDefault()',
                    },
                }]
        }] });

const BRN_SWITCH_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnSwitchComponent),
    multi: true,
};
const CONTAINER_POST_FIX = '-switch';
let uniqueIdCounter = 0;
class BrnSwitchComponent {
    _destroyRef = inject(DestroyRef);
    _renderer = inject(Renderer2);
    _isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    _elementRef = inject(ElementRef);
    _focusMonitor = inject(FocusMonitor);
    _cdr = inject(ChangeDetectorRef);
    _document = inject(DOCUMENT);
    focusVisible = signal(false);
    focused = signal(false);
    /**
     * Whether the switch is checked.
     * Can be bound with [(checked)]
     */
    checked = model(false);
    /**
     * Sets the ID on the switch.
     * When provided, the inner button gets this ID without the '-switch' suffix.
     */
    id = input(uniqueIdCounter++ + '');
    /**
     * Sets the name on the switch.
     * When provided, the inner button gets this name without a '-switch' suffix.
     */
    name = input(null);
    /**
     * Sets class set on the inner button
     */
    class = input(null);
    /**
     * Sets the aria-label attribute for accessibility.
     */
    ariaLabel = input(null, { alias: 'aria-label' });
    /**
     * Sets the aria-labelledby attribute for accessibility.
     */
    ariaLabelledby = input(null, { alias: 'aria-labelledby' });
    mutableAriaLabelledby = linkedSignal(() => this.ariaLabelledby());
    /**
     * Sets the aria-describedby attribute for accessibility.
     */
    ariaDescribedby = input(null, { alias: 'aria-describedby' });
    /**
     * Whether the switch is required in a form.
     */
    required = input(false, { transform: booleanAttribute });
    /**
     * Whether the switch is disabled.
     */
    disabled = input(false, {
        transform: booleanAttribute,
    });
    /**
     * tabIndex of the switch.
     */
    tabIndex = input(0);
    /**
     * Event emitted when the switch value changes.
     */
    changed = output();
    /**
     * Event emitted when the switch is blurred (loses focus).
     */
    touched = output();
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onChange = () => { };
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onTouched = () => { };
    switch = viewChild.required('switch');
    state = computed(() => {
        const name = this.name();
        const id = this.id();
        return {
            disabled: signal(this.disabled()),
            name: name ? name + CONTAINER_POST_FIX : null,
            id: id ? id + CONTAINER_POST_FIX : null,
        };
    });
    constructor() {
        effect(() => {
            const state = this.state();
            const isDisabled = state.disabled();
            if (!this._elementRef.nativeElement || !this._isBrowser)
                return;
            const newLabelId = state.id + '-label';
            const switchButtonId = this.getSwitchButtonId(state.id);
            const labelElement = this._elementRef.nativeElement.closest('label') ??
                this._document.querySelector(`label[for="${switchButtonId}"]`);
            if (!labelElement)
                return;
            const existingLabelId = labelElement.id;
            this._renderer.setAttribute(labelElement, 'data-disabled', isDisabled ? 'true' : 'false');
            this.mutableAriaLabelledby.set(existingLabelId || newLabelId);
            if (!existingLabelId || existingLabelId.length === 0) {
                this._renderer.setAttribute(labelElement, 'id', newLabelId);
            }
        });
    }
    toggle() {
        if (this.state().disabled())
            return;
        this.checked.update((checked) => !checked);
        this._onChange(this.checked());
        this.changed.emit(this.checked());
    }
    ngAfterContentInit() {
        this._focusMonitor
            .monitor(this._elementRef, true)
            .pipe(takeUntilDestroyed(this._destroyRef))
            .subscribe((focusOrigin) => {
            if (focusOrigin)
                this.focused.set(true);
            if (focusOrigin === 'keyboard' || focusOrigin === 'program') {
                this.focusVisible.set(true);
                this._cdr.markForCheck();
            }
            if (!focusOrigin) {
                // When a focused element becomes disabled, the browser *immediately* fires a blur event.
                // Angular does not expect events to be raised during change detection, so any state
                // change (such as a form control's ng-touched) will cause a changed-after-checked error.
                // See https://github.com/angular/angular/issues/17793. To work around this, we defer
                // telling the form control it has been touched until the next tick.
                Promise.resolve().then(() => {
                    this.focusVisible.set(false);
                    this.focused.set(false);
                    this._onTouched();
                    this.touched.emit();
                    this._cdr.markForCheck();
                });
            }
        });
        if (!this.switch())
            return;
        this.switch().nativeElement.value = this.checked() ? 'on' : 'off';
        this.switch().nativeElement.dispatchEvent(new Event('change'));
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    /** We intercept the id passed to the wrapper component and pass it to the underlying button switch control **/
    getSwitchButtonId(idPassedToContainer) {
        return idPassedToContainer ? idPassedToContainer.replace(CONTAINER_POST_FIX, '') : null;
    }
    writeValue(value) {
        this.checked.set(Boolean(value));
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /** Implemented as a part of ControlValueAccessor. */
    setDisabledState(isDisabled) {
        this.state().disabled.set(isDisabled);
        this._cdr.markForCheck();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "19.2.8", type: BrnSwitchComponent, isStandalone: true, selector: "brn-switch", inputs: { checked: { classPropertyName: "checked", publicName: "checked", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, name: { classPropertyName: "name", publicName: "name", isSignal: true, isRequired: false, transformFunction: null }, class: { classPropertyName: "class", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedby: { classPropertyName: "ariaDescribedby", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, tabIndex: { classPropertyName: "tabIndex", publicName: "tabIndex", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checked: "checkedChange", changed: "changed", touched: "touched" }, host: { properties: { "style": "{display: \"contents\"}", "attr.id": "state().id", "attr.name": "state().name", "attr.aria-labelledby": "null", "attr.aria-label": "null", "attr.aria-describedby": "null", "attr.data-state": "checked() ? \"checked\" : \"unchecked\"", "attr.data-focus-visible": "focusVisible()", "attr.data-focus": "focused()", "attr.data-disabled": "state().disabled()" } }, providers: [BRN_SWITCH_VALUE_ACCESSOR], viewQueries: [{ propertyName: "switch", first: true, predicate: ["switch"], descendants: true, isSignal: true }], ngImport: i0, template: `
		<button
			#switch
			role="switch"
			type="button"
			[class]="class()"
			[id]="getSwitchButtonId(state().id) ?? ''"
			[name]="getSwitchButtonId(state().name) ?? ''"
			[value]="checked() ? 'on' : 'off'"
			[attr.aria-checked]="checked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.data-state]="checked() ? 'checked' : 'unchecked'"
			[attr.data-focus-visible]="focusVisible()"
			[attr.data-focus]="focused()"
			[attr.data-disabled]="state().disabled()"
			[disabled]="state().disabled()"
			[tabIndex]="tabIndex()"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content select="brn-switch-thumb" />
		</button>
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-switch',
                    template: `
		<button
			#switch
			role="switch"
			type="button"
			[class]="class()"
			[id]="getSwitchButtonId(state().id) ?? ''"
			[name]="getSwitchButtonId(state().name) ?? ''"
			[value]="checked() ? 'on' : 'off'"
			[attr.aria-checked]="checked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.data-state]="checked() ? 'checked' : 'unchecked'"
			[attr.data-focus-visible]="focusVisible()"
			[attr.data-focus]="focused()"
			[attr.data-disabled]="state().disabled()"
			[disabled]="state().disabled()"
			[tabIndex]="tabIndex()"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content select="brn-switch-thumb" />
		</button>
	`,
                    host: {
                        '[style]': '{display: "contents"}',
                        '[attr.id]': 'state().id',
                        '[attr.name]': 'state().name',
                        '[attr.aria-labelledby]': 'null',
                        '[attr.aria-label]': 'null',
                        '[attr.aria-describedby]': 'null',
                        '[attr.data-state]': 'checked() ? "checked" : "unchecked"',
                        '[attr.data-focus-visible]': 'focusVisible()',
                        '[attr.data-focus]': 'focused()',
                        '[attr.data-disabled]': 'state().disabled()',
                    },
                    providers: [BRN_SWITCH_VALUE_ACCESSOR],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], ctorParameters: () => [] });

const BrnSwitchImports = [BrnSwitchComponent, BrnSwitchThumbComponent];
class BrnSwitchModule {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    /** @nocollapse */ static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchModule, imports: [BrnSwitchComponent, BrnSwitchThumbComponent], exports: [BrnSwitchComponent, BrnSwitchThumbComponent] });
    /** @nocollapse */ static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnSwitchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnSwitchImports],
                    exports: [...BrnSwitchImports],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_SWITCH_VALUE_ACCESSOR, BrnSwitchComponent, BrnSwitchImports, BrnSwitchModule, BrnSwitchThumbComponent };
