// @spartan-ng/brain/fesm2022/spartan-ng-brain-checkbox.mjs
import * as i0 from '@angular/core';
import { forwardRef, booleanAttribute, inject, Renderer2, ElementRef, PLATFORM_ID, signal, model, computed, input, viewChild, output, effect, HostListener, ViewEncapsulation, ChangeDetectionStrategy, Component, NgModule } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { isPlatformBrowser, NgStyle } from '@angular/common';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

const BRN_CHECKBOX_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnCheckboxComponent),
    multi: true,
};
function indeterminateBooleanAttribute(value) {
    if (value === 'indeterminate')
        return 'indeterminate';
    return booleanAttribute(value);
}
const CONTAINER_POST_FIX = '-checkbox';
class BrnCheckboxComponent {
    _renderer = inject(Renderer2);
    _elementRef = inject(ElementRef);
    _focusMonitor = inject(FocusMonitor);
    _isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    _focusVisible = signal(false);
    focusVisible = this._focusVisible.asReadonly();
    _focused = signal(false);
    focused = this._focused.asReadonly();
    checked = model(false);
    isChecked = this.checked.asReadonly();
    _dataState = computed(() => {
        const checked = this.checked();
        if (checked === 'indeterminate')
            return 'indeterminate';
        return checked ? 'checked' : 'unchecked';
    });
    _ariaChecked = computed(() => {
        const checked = this.checked();
        if (checked === 'indeterminate')
            return 'mixed';
        return checked ? 'true' : 'false';
    });
    _value = computed(() => {
        const checked = this.checked();
        if (checked === 'indeterminate')
            return '';
        return checked ? 'on' : 'off';
    });
    /** Used to set the id on the underlying input element. */
    id = input(null);
    hostId = computed(() => (this.id() ? this.id() + CONTAINER_POST_FIX : null));
    /** Used to set the name attribute on the underlying input element. */
    name = input(null);
    hostName = computed(() => (this.name() ? this.name() + CONTAINER_POST_FIX : null));
    /** Used to set the aria-label attribute on the underlying input element. */
    ariaLabel = input(null, { alias: 'aria-label' });
    /** Used to set the aria-labelledby attribute on the underlying input element. */
    ariaLabelledby = input(null, { alias: 'aria-labelledby' });
    ariaDescribedby = input(null, { alias: 'aria-describedby' });
    required = input(false, { transform: booleanAttribute });
    disabled = input(false, { transform: booleanAttribute });
    state = computed(() => ({
        disabled: signal(this.disabled()),
    }));
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onChange = () => { };
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onTouched = () => { };
    checkbox = viewChild.required('checkBox');
    changed = output();
    constructor() {
        effect(() => {
            const parent = this._renderer.parentNode(this._elementRef.nativeElement);
            if (!parent)
                return;
            // check if parent is a label and assume it is for this checkbox
            if (parent?.tagName === 'LABEL') {
                this._renderer.setAttribute(parent, 'data-disabled', this.state().disabled() ? 'true' : 'false');
                return;
            }
            if (!this._isBrowser)
                return;
            const label = parent?.querySelector(`label[for="${this.id()}"]`);
            if (!label)
                return;
            this._renderer.setAttribute(label, 'data-disabled', this.state().disabled() ? 'true' : 'false');
        });
    }
    toggle(event) {
        if (this.state().disabled())
            return;
        event.preventDefault();
        const previousChecked = this.checked();
        this.checked.set(previousChecked === 'indeterminate' ? true : !previousChecked);
        this._onChange(!previousChecked);
        this.changed.emit(!previousChecked);
    }
    ngAfterContentInit() {
        this._focusMonitor.monitor(this._elementRef, true).subscribe((focusOrigin) => {
            if (focusOrigin)
                this._focused.set(true);
            if (focusOrigin === 'keyboard' || focusOrigin === 'program') {
                this._focusVisible.set(true);
            }
            if (!focusOrigin) {
                // When a focused element becomes disabled, the browser *immediately* fires a blur event.
                // Angular does not expect events to be raised during change detection, so any state
                // change (such as a form control's ng-touched) will cause a changed-after-checked error.
                // See https://github.com/angular/angular/issues/17793. To work around this, we defer
                // telling the form control it has been touched until the next tick.
                Promise.resolve().then(() => {
                    this._focusVisible.set(false);
                    this._focused.set(false);
                    this._onTouched();
                });
            }
        });
        this.checkbox().nativeElement.indeterminate = this.checked() === 'indeterminate';
        if (this.checkbox().nativeElement.indeterminate) {
            this.checkbox().nativeElement.value = 'indeterminate';
        }
        else {
            this.checkbox().nativeElement.value = this.checked() ? 'on' : 'off';
        }
        this.checkbox().nativeElement.dispatchEvent(new Event('change'));
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    writeValue(value) {
        if (value === 'indeterminate') {
            this.checked.set('indeterminate');
        }
        else {
            this.checked.set(!!value);
        }
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /** Implemented as a part of ControlValueAccessor. */
    setDisabledState(isDisabled) {
        this.state().disabled.set(isDisabled);
    }
    /**
     * If the space key is pressed, prevent the default action to stop the page from scrolling.
     */
    preventScrolling(event) {
        event.preventDefault();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "19.2.8", type: BrnCheckboxComponent, isStandalone: true, selector: "brn-checkbox", inputs: { checked: { classPropertyName: "checked", publicName: "checked", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, name: { classPropertyName: "name", publicName: "name", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedby: { classPropertyName: "ariaDescribedby", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checked: "checkedChange", changed: "changed" }, host: { listeners: { "click": "toggle($event)", "keyup.space": "toggle($event)", "keyup.enter": "toggle($event)", "keydown.space": "preventScrolling($event)" }, properties: { "attr.tabindex": "state().disabled() ? \"-1\" : \"0\"", "attr.data-state": "_dataState()", "attr.data-focus-visible": "focusVisible()", "attr.data-focus": "focused()", "attr.data-disabled": "state().disabled()", "attr.aria-labelledby": "null", "attr.aria-label": "null", "attr.aria-describedby": "null", "attr.id": "hostId()", "attr.name": "hostName()" } }, providers: [BRN_CHECKBOX_VALUE_ACCESSOR], viewQueries: [{ propertyName: "checkbox", first: true, predicate: ["checkBox"], descendants: true, isSignal: true }], ngImport: i0, template: `
		<input
			#checkBox
			tabindex="-1"
			type="checkbox"
			role="checkbox"
			[ngStyle]="{
				position: 'absolute',
				width: '1px',
				height: '1px',
				padding: '0',
				margin: '-1px',
				overflow: 'hidden',
				clip: 'rect(0, 0, 0, 0)',
				whiteSpace: 'nowrap',
				borderWidth: '0',
			}"
			[id]="id() ?? ''"
			[name]="name() ?? ''"
			[value]="_value()"
			[checked]="isChecked()"
			[required]="required()"
			[attr.aria-label]="ariaLabel()"
			[attr.aria-labelledby]="ariaLabelledby()"
			[attr.aria-describedby]="ariaDescribedby()"
			[attr.aria-required]="required() || null"
			[attr.aria-checked]="_ariaChecked()"
		/>
		<ng-content />
	`, isInline: true, dependencies: [{ kind: "directive", type: NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-checkbox',
                    imports: [NgStyle],
                    template: `
		<input
			#checkBox
			tabindex="-1"
			type="checkbox"
			role="checkbox"
			[ngStyle]="{
				position: 'absolute',
				width: '1px',
				height: '1px',
				padding: '0',
				margin: '-1px',
				overflow: 'hidden',
				clip: 'rect(0, 0, 0, 0)',
				whiteSpace: 'nowrap',
				borderWidth: '0',
			}"
			[id]="id() ?? ''"
			[name]="name() ?? ''"
			[value]="_value()"
			[checked]="isChecked()"
			[required]="required()"
			[attr.aria-label]="ariaLabel()"
			[attr.aria-labelledby]="ariaLabelledby()"
			[attr.aria-describedby]="ariaDescribedby()"
			[attr.aria-required]="required() || null"
			[attr.aria-checked]="_ariaChecked()"
		/>
		<ng-content />
	`,
                    host: {
                        '[attr.tabindex]': 'state().disabled() ? "-1" : "0"',
                        '[attr.data-state]': '_dataState()',
                        '[attr.data-focus-visible]': 'focusVisible()',
                        '[attr.data-focus]': 'focused()',
                        '[attr.data-disabled]': 'state().disabled()',
                        '[attr.aria-labelledby]': 'null',
                        '[attr.aria-label]': 'null',
                        '[attr.aria-describedby]': 'null',
                        '[attr.id]': 'hostId()',
                        '[attr.name]': 'hostName()',
                    },
                    providers: [BRN_CHECKBOX_VALUE_ACCESSOR],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], ctorParameters: () => [], propDecorators: { toggle: [{
                type: HostListener,
                args: ['click', ['$event']]
            }, {
                type: HostListener,
                args: ['keyup.space', ['$event']]
            }, {
                type: HostListener,
                args: ['keyup.enter', ['$event']]
            }], preventScrolling: [{
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }] } });

const BrnCheckboxImports = [BrnCheckboxComponent];
class BrnCheckboxModule {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    /** @nocollapse */ static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxModule, imports: [BrnCheckboxComponent], exports: [BrnCheckboxComponent] });
    /** @nocollapse */ static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.8", ngImport: i0, type: BrnCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...BrnCheckboxImports],
                    exports: [...BrnCheckboxImports],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_CHECKBOX_VALUE_ACCESSOR, BrnCheckboxComponent, BrnCheckboxImports, BrnCheckboxModule, indeterminateBooleanAttribute };
