// @nx/rspack/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eslintPluginReactHooksVersion = exports.eslintPluginReactVersion = exports.eslintPluginJsxA11yVersion = exports.eslintPluginImportVersion = exports.stylusVersion = exports.lessVersion = exports.nestjsMicroservicesVersion = exports.nestjsPlatformExpressVersion = exports.nestjsCoreVersion = exports.nestjsCommonVersion = exports.reactRefreshVersion = exports.sassEmbeddedVersion = exports.sassLoaderVersion = exports.lessLoaderVersion = exports.rspackPluginReactRefreshVersion = exports.rspackDevServerVersion = exports.rspackCoreVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.rspackCoreVersion = '1.2.2';
exports.rspackDevServerVersion = '1.0.9';
exports.rspackPluginReactRefreshVersion = '^1.0.0';
exports.lessLoaderVersion = '~11.1.3';
exports.sassLoaderVersion = '^16.0.4';
exports.sassEmbeddedVersion = '^1.83.4';
exports.reactRefreshVersion = '~0.14.0';
exports.nestjsCommonVersion = '~9.0.0';
exports.nestjsCoreVersion = '~9.0.0';
exports.nestjsPlatformExpressVersion = '~9.0.0';
exports.nestjsMicroservicesVersion = '~9.0.0';
exports.lessVersion = '4.1.3';
exports.stylusVersion = '^0.55.0';
exports.eslintPluginImportVersion = '2.27.5';
exports.eslintPluginJsxA11yVersion = '6.7.1';
exports.eslintPluginReactVersion = '7.32.2';
exports.eslintPluginReactHooksVersion = '4.6.0';
